<a href="https://colab.research.google.com/github/jiyoung-99/boiler-plate-node/blob/master/n112a_feature_engineering_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://user-images.githubusercontent.com/6457691/89746244-a6e2a200-daf3-11ea-9314-7863d544d73c.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 1 / NOTE 2*

# 📝 Assignment


## 1. Feature Engineering

- `KT&G` 기업 데이터를 계속 사용합니다.

- 각 분기별 영업이익률을 직접 계산해보세요. 

- 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요. 단 여기서 점유율은 일반담배 기준이며 18년:62%, 19년 63.5% 로 계산합니다. 시장의 규모는 18년 15조, 19년 16조 라고 가정합니다. 

참조 : 
<http://it.chosun.com/site/data/html_dir/2020/02/13/2020021302434.html>


In [1]:
# KT&G 기업데이터 불러오기, 숫자는 억 단위
import pandas as pd

df = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g.csv')

df

분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

In [2]:
# 각 분기별 영업이익률을 직접 계산해보세요.

df.dtypes

분기             object
매출액            object
영업이익           object
영업이익(발표기준)     object
세전계속사업이익       object
당기순이익          object
당기순이익(지배)      object
당기순이익(비지배)      int64
자산총계           object
부채총계           object
자본총계           object
자본총계(지배)       object
자본총계(비지배)     float64
자본금            object
영업활동현금흐름       object
투자활동현금흐름       object
재무활동현금흐름       object
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)         object
PER(배)        float64
dtype: object

In [3]:
# 계산 전 모두 숫자로 바꿔준다.
# 숫자로 바꿔주는 함수 선언, replce 쓰지 않음
def to_int(is_string):
    return int(''.join(is_string.split(',')))

In [4]:
# 영업이익율에 필요한 영업이익, 매출액 컬럼 숫자로 변경
df['매출액_int'] = df['매출액'].apply(to_int)
df['영업이익_int'] = df['영업이익'].apply(to_int)
df.dtypes

분기             object
매출액            object
영업이익           object
영업이익(발표기준)     object
세전계속사업이익       object
당기순이익          object
당기순이익(지배)      object
당기순이익(비지배)      int64
자산총계           object
부채총계           object
자본총계           object
자본총계(지배)       object
자본총계(비지배)     float64
자본금            object
영업활동현금흐름       object
투자활동현금흐름       object
재무활동현금흐름       object
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)         object
PER(배)        float64
매출액_int         int64
영업이익_int        int64
dtype: object

In [5]:
# 영업이익/매출액 X 100 = 영업이익률
# round 함수로 소숫점 자리 2째자리까지 표현
#################1번 답##################
df['영업이익직접'] = df['영업이익_int'] / df['매출액_int'] * 100
df['영업이익직접'] = round(df['영업이익직접'], 2)
df

분기     매출액   영업이익 영업이익(발표기준)  ... PER(배) 매출액_int 영업이익_int  영업이익직접
0   19/6  12,578  3,991      3,991  ...  13.77   12578     3991   31.73
1   19/9  13,222  3,825      3,825  ...  13.87   13222     3825   28.93
2  19/12  11,982  2,523      2,523  ...  12.42   11982     2523   21.06
3   20/3  11,784  3,150      3,150  ...   9.71   11784     3150   26.73
4   20/6  13,188  3,947      3,947  ...  10.46   13188     3947   29.93

[5 rows x 28 columns]

In [ ]:
#18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요. 단 여기서 점유율은 일반담배 기준이며 18년:62%, 19년 63.5% 로 계산합니다. 
# 시장의 규모는 18년 15조, 19년 16조 라고 가정합니다.
# 시가총액을 TAM에 해당 기업의 시장점유율을 곱한 값으로 나눠 PDR을 산출한다. 
# 예를 들어 시가총액 5조원 규모의 A업체가 100조원의 시장 규모를 가진 산업 내에서 10%의 시장점유율을 갖고 있다면 PDR은 0.5다.

# 2018년 : 13.8조(시총)/15조*0.62
# 2019년 : 12조/16조*0.635

In [6]:
#2018년  138,000 / (150,000*0.62)
#2019년  120,000/160,000*0.635

#변수 각 년도별 시가총액
mc_2018 = 138000
mc_2019 = 120000
#TAM 변수
tam_2018 = 150000
tam_2019 = 160000
# 시장점유율 변수
ms_2018 = 0.62
ms_2019 = 0.635



In [7]:
pdr_dict = {'연도':['2018','2019']
            ,'시가총액':[mc_2018, mc_2019]
            ,'TAM':[tam_2018,tam_2019]
            ,'시장점유율':[ms_2018,ms_2019]
            }


In [8]:
#데이터프레임만들기
df_pdr = pd.DataFrame(pdr_dict)
df_pdr.head()
df_pdr['PDR'] = df_pdr['시가총액']/(df_pdr['TAM']*df_pdr['시장점유율'])
#######################1.2번답#######################
# pdr 추가해서 dataframe 출력
df_pdr.head()

연도    시가총액     TAM  시장점유율       PDR
0  2018  138000  150000  0.620  1.483871
1  2019  120000  160000  0.635  1.181102

## 2. String Manipulation

- replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하세요. 결과값은 12578이 나와야합니다.


In [9]:

################# 2번 답 #################
not_replace = int(''.join(df['매출액'][0].split(',')))
print(not_replace)
type(not_replace)

12578


int

## 3. Apply
- 데이터셋의 분기를 제외한 모든 부분에 대해서 `describe`를 통해 summary statistics 를 확인하세요.
- 단 결측치는 0으로 처리합니다.

In [10]:
# 다시 선언
df_apply = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g.csv')

In [11]:
# 결측치 0으로 처리
df_apply = df_apply.fillna(0)
df_apply.head()

분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

In [12]:
# dataframe 에서 분기 컬럼 삭제
df_apply = df_apply.drop('분기', axis=1)
df_apply


매출액   영업이익 영업이익(발표기준) 세전계속사업이익  ...   부채비율   자본유보율  EPS(원) PER(배)
0  12,578  3,991      3,991    4,452  ...  31.43  803.22   2,351  13.77
1  13,222  3,825      3,825    4,503  ...  24.17  836.86   2,337  13.87
2  11,982  2,523      2,523    1,794  ...  23.04  850.82     873  12.42
3  11,784  3,150      3,150    4,156  ...  29.69  823.16   2,134   9.71
4  13,188  3,947      3,947    4,020  ...  26.91  837.40   2,135  10.46

[5 rows x 24 columns]

In [13]:
df_apply.dtypes

매출액            object
영업이익           object
영업이익(발표기준)     object
세전계속사업이익       object
당기순이익          object
당기순이익(지배)      object
당기순이익(비지배)      int64
자산총계           object
부채총계           object
자본총계           object
자본총계(지배)       object
자본총계(비지배)     float64
자본금            object
영업활동현금흐름       object
투자활동현금흐름       object
재무활동현금흐름       object
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)         object
PER(배)        float64
dtype: object

In [14]:
# object 형은 모두 int 형으로 변경
df_apply['매출액'] = df_apply['매출액'].apply(to_int)
df_apply['영업이익'] = df_apply['영업이익'].apply(to_int)
df_apply['영업이익(발표기준)'] = df_apply['영업이익(발표기준)'].apply(to_int)
df_apply['세전계속사업이익'] = df_apply['세전계속사업이익'].apply(to_int)
df_apply['당기순이익'] = df_apply['당기순이익'].apply(to_int)
df_apply['당기순이익(지배)'] = df_apply['당기순이익(지배)'].apply(to_int)
df_apply['자산총계'] = df_apply['자산총계'].apply(to_int)
df_apply['부채총계'] = df_apply['부채총계'].apply(to_int)
df_apply['자본총계'] = df_apply['자본총계'].apply(to_int)
df_apply['자본총계(지배)'] = df_apply['자본총계(지배)'].apply(to_int)
df_apply['자본금'] = df_apply['자본금'].apply(to_int)
df_apply['영업활동현금흐름'] = df_apply['영업활동현금흐름'].apply(to_int)
df_apply['투자활동현금흐름'] = df_apply['투자활동현금흐름'].apply(to_int)
df_apply['재무활동현금흐름'] = df_apply['재무활동현금흐름'].apply(to_int)
df_apply['EPS(원)'] = df_apply['EPS(원)'].apply(to_int)
df_apply.dtypes

매출액             int64
영업이익            int64
영업이익(발표기준)      int64
세전계속사업이익        int64
당기순이익           int64
당기순이익(지배)       int64
당기순이익(비지배)      int64
자산총계            int64
부채총계            int64
자본총계            int64
자본총계(지배)        int64
자본총계(비지배)     float64
자본금             int64
영업활동현금흐름        int64
투자활동현금흐름        int64
재무활동현금흐름        int64
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)          int64
PER(배)        float64
dtype: object

In [15]:
# 데이터셋의 분기를 제외한 모든 부분에 대해서 describe를 통해 summary statistics 를 확인하세요.
##################3번답##################
df_apply.describe()

매출액         영업이익  ...       EPS(원)     PER(배)
count      5.000000     5.000000  ...     5.000000   5.000000
mean   12550.800000  3487.200000  ...  1966.000000  12.046000
std      664.979097   636.931079  ...   619.939513   1.898112
min    11784.000000  2523.000000  ...   873.000000   9.710000
25%    11982.000000  3150.000000  ...  2134.000000  10.460000
50%    12578.000000  3825.000000  ...  2135.000000  12.420000
75%    13188.000000  3947.000000  ...  2337.000000  13.770000
max    13222.000000  3991.000000  ...  2351.000000  13.870000

[8 rows x 24 columns]

## 🔥 도전과제



### NA Value Handling

- 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
- 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요.



In [80]:
import numpy as np


# 다시 데이터 받아오기
df_plus = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g.csv')
# 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
df_plus.head()


분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

In [85]:
#결측값으로 변경
##############도전과제1.1번 답 ###############3
df_plus.iloc[2,5] = np.NaN
df_plus.head()

분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

In [86]:
# 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요
# mean imputation 은 결측치에 평균 값 넣기###############
df_plus['당기순이익(비지배)'] = df_plus['당기순이익(비지배)'].fillna(df_plus['당기순이익(비지배)'].mean())
df_plus.head()

분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

### Feature Engineering

Relative Perfomance 라는 새로운 feature를 계산하세요.

이는 최근 1년치 매출액의 평균값을 기준으로 

- 10% 이상 -> `S`
- 5% 이상 -> `A`
- -5 ~ 5% -> `B`
- -5%이하 -> `C`
- 10%이하 -> `D`

라는 값을 갖는 feature입니다.

20년도 2분기에 해당하는 결과값은 `A`가 나와야합니다.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하세요.

In [56]:
#우선 평균 먼저 구한다.

df_sale = df_plus[['분기','매출액']]
df_sale
# 최근 1녀간 평균 매기는데 필요없는 0번째 인덱스 삭제
df_sale_drop = df_sale.drop(index=0, axis=1)
# 숫자로 바꿔준다.
df_sale_drop['매출액'] = df_sale_drop['매출액'].apply(to_int)


In [57]:
df_sale_drop.head()

분기    매출액
1   19/9  13222
2  19/12  11982
3   20/3  11784
4   20/6  13188

In [59]:
# mean()을 써서 평균 구한다.
sale_mean = df_sale_drop.mean()
print(sale_mean)

매출액    12544.0
dtype: float64


In [61]:
# 평균 12544 를 변수로 선언
mean_value = 12544

In [66]:
# 10% 이상 -> S
# 5% 이상 -> A
# -5 ~ 5% -> B
# -5%이하 -> C
# 10%이하 -> D

print('S : ', round(mean_value*1.1, 1), '% 이상')
print('A : ', mean_value*1.05, '% 이상')
print('B : ', mean_value*0.95, '% 이상 ~ ', mean_value*1.05, '% 미만')
print('C : ', mean_value*0.95, '% 이하')
print('D : ', mean_value*0.9, '% 이하')

S :  13798.4 % 이상
A :  13171.2 % 이상
B :  11916.8 % 이상 ~  13171.2 % 미만
C :  11916.8 % 이하
D :  11289.6 % 이하
